lux-lua/src/loader.rs

```diff
@@ -0,0 +1,190 @@
+    lockfile::{LocalPackageId, LocalPackageLockType, Lockfile},
+    package::PackageReq,
+    project::Project,
+};
+use mlua::prelude::*;
+use path_absolutize::Absolutize;
+
+const fn lua_version() -> &'static str {
+    if cfg!(feature = "lua51") {
+        "5.1"
+    } else if cfg!(feature = "lua52") {
+        "5.2"
+    } else if cfg!(feature = "lua53") {
+        "5.3"
+    } else if cfg!(feature = "lua54") {
+        "5.4"
+    } else {
+        unreachable!()
+    }
+}
+
+fn current_file(lua: &Lua) -> String {
+    lua.inspect_stack(2)
+        .unwrap()
+        .source()
+        .short_src
+        .unwrap()
+        .to_string()
+}
+
+fn load_file(lua: &Lua, module: String, path: PathBuf) -> mlua::Result<mlua::Function> {
+    lua.create_function(move |lua, ()| {
+        // We need to handle 3 possible cases:
+        // - `src/?.lua`
+        // - `src/?/init.lua`
+        // - `src/?.so`
+
+        let module_path = module.replace('.', "/");
+
+        if path
+            .join("src")
+            .join(format!("{}.lua", module_path))
+            .exists()
+        {
+            lua.load("dofile")
+                .call::<()>(path.join("src").join(format!("{}.lua", module_path)))
+        } else if path
+            .join("src")
+            .join(&module_path)
+            .join("init.lua")
+            .exists()
+        {
+            lua.load("dofile")
+                .call::<()>(path.join("src").join(&module_path).join("init.lua"))
+        } else if path
+            .join("lib")
+            .join(format!("{}.so", module_path))
+            .exists()
+        {
+            lua.load("dofile")
+                .call::<()>(path.join("lib").join(format!("{}.so", module_path)))
+        } else {
+            Err(mlua::Error::RuntimeError(format!(
+                "module not found: {}",
+                module
+            )))
+        }
+    })
+}
+
+pub fn load_loader(lua: &Lua) -> mlua::Result<()> {
+    let globals = lua.globals();
+    let package: LuaTable = globals.get("package")?;
+    #[cfg(feature = "lua51")]
+    let loaders: LuaTable = package.get("loaders")?;
+    #[cfg(not(feature = "lua51"))]
+    let loaders: LuaTable = package.get("searchers")?;
+    loaders.raw_insert(1, lua.create_function(loader)?)?;
+
+    Ok(())
+}
+
+pub fn loader(lua: &Lua, module: String) -> mlua::Result<Option<mlua::Function>> {
+    let module_root = match module.split('.').next() {
+        Some(module) => module,
+        None => return Ok(None),
+    };
+    let mut current_file = match current_file(lua).as_str() {
+        current_file => PathBuf::from(current_file),
+        "stdin" => return Ok(None),
+    };
+    let current_file = current_file.absolutize().into_lua_err()?;
+
+    match lets_find_up::find_up_with(
+        "lux.lock",
+        lets_find_up::FindUpOptions {
+            cwd: current_file.parent().unwrap(),
+            kind: lets_find_up::FindUpKind::File,
+        },
+    )
+    .into_lua_err()?
```

Comment:
question: could searching for the lockfile on every `require` be expensive?
Perhaps we should cache the results for `current_file`. I suppose since we're in the Lua space, we would have to use a Lua global.

Reply:
This implementation was a proof of concept. I'll definitely be spending time optimizing this further with a caching mechanism and with some heuristics to prevent having to traverse the filesystem itself.

Reply:
New implementation completely omits any filesystem reads for lookup and simply loops over the path in-memory to try and find the correct path :)